File: backend/room/src/commonMain/kotlin/com/rafaelrain/message/backend/room/Room.kt
package com.rafaelrain.message.backend.room

import com.rafaelrain.chatmessage.common.extensions.now
import com.rafaelrain.chatmessage.common.packet.ClientMessageSessionPacket
import com.rafaelrain.chatmessage.common.packet.ClientMessageSessionPacketType
import com.rafaelrain.chatmessage.common.packet.ServerMessageSessionPacket
import com.rafaelrain.chatmessage.common.packet.ServerMessageSessionPacketType
import com.rafaelrain.message.backend.room.connection.PacketClient
import com.rafaelrain.message.backend.room.message.ChatMessage
import com.rafaelrain.message.backend.room.session.MessageSession
import kotlinx.datetime.LocalDateTime

data class Room<C : PacketClient>(
    val name: String,
    val sessions: MutableMap<String, MessageSession<C>> = mutableMapOf(),
    val messages: MutableList<ChatMessage> = mutableListOf(),
    val createdAt: LocalDateTime = LocalDateTime.now(),
) {
    suspend fun addNewSession(session: MessageSession<C>) {
        sessions.remove(session.name)
        sessions[session.name] = session

        notifyJoin(session)
        sendMessageHistory(session)
    }

    suspend fun handleLeave(session: MessageSession<C>) {
        sessions.remove(session.name)

        broadcastPacket(
            ServerMessageSessionPacket.leave(roomName = this.name, sessionName = session.name),
        )
    }

    suspend fun handleClientPacket(
        session: MessageSession<C>,
        packet: ClientMessageSessionPacket,
    ) {
        when (packet.type) {
            ClientMessageSessionPacketType.MESSAGE -> handleMessage(session, packet.message.orEmpty())
        }
    }

    private suspend fun notifyJoin(joinedUser: MessageSession<C>) {
        val packet =
            ServerMessageSessionPacket(
                type = ServerMessageSessionPacketType.JOINED,
                roomName = this.name,
                senderName = joinedUser.name,
            )
        broadcastPacket(packet)
    }

    private suspend fun sendMessageHistory(session: MessageSession<C>) {
        this.messages.asSequence()
            .sortedBy { message -> message.sentAt }
            .map { message -> message.toServerPacket() }
            .forEach { packet -> sendPacketOrDisconnect(session, packet) }
    }

    private suspend fun broadcastPacket(packet: ServerMessageSessionPacket) {
        sessions.forEach { (_, session) ->
            sendPacketOrDisconnect(session, packet)
        }
    }

    private suspend fun sendPacketOrDisconnect(
        session: MessageSession<C>,
        packet: ServerMessageSessionPacket,
    ) {
        runCatching {
            session.sendPacket(packet)
        }.onFailure {
            handleLeave(session)
            runCatching { session.disconnect() }
        }
    }

    private suspend fun handleMessage(
        session: MessageSession<C>,
        message: String,
    ) {
        val chatMessage = ChatMessage(sender = session.name, message = message, roomName = this.name)

        broadcastPacket(chatMessage.toServerPacket())
        messages.add(chatMessage)
    }
}

File: backend-rust/src/domain/rooms/room.rs
use crate::domain::packets::{ClientPacket, ServerPacket};
use crate::domain::rooms::message::ChatMessage;
use crate::domain::rooms::session::MessageSession;
use crate::error::Error::BlankMessageError;
use anyhow::Result;
use jiff::Zoned;

use futures::future::join_all;
use std::collections::HashMap;

pub struct Room {
    pub name: String,
    pub sessions: HashMap<String, MessageSession>,
    pub messages: Vec<ChatMessage>,
    pub created_at: Zoned,
}

impl Room {
    pub fn new(name: String) -> Room {
        Room {
            name,
            sessions: HashMap::new(),
            messages: Vec::new(),
            created_at: Zoned::now(),
        }
    }

    pub async fn add_new_session(&mut self, mut session: MessageSession) -> Result<()> {
        self.sessions.remove(&session.name);

        self.notify_join(&session).await?;
        self.send_message_history(&mut session).await?;

        self.sessions.insert(session.name.clone(), session);

        Ok(())
    }

    pub async fn handle_leave(&mut self, sender_name: String) -> Result<()> {
        self.sessions.remove(&sender_name);

        let leave_packet =
            ServerPacket::create_user_left(self.name.to_string(), sender_name, Zoned::now());

        self.broadcast_packet(leave_packet).await?;
        Ok(())
    }

    pub async fn handle_client_packet(
        &mut self,
        sender_name: String,
        client_packet: ClientPacket,
    ) -> Result<()> {
        match client_packet {
            ClientPacket::Message { message, .. } => {
                self.handle_message(sender_name, message).await?
            }
            ClientPacket::Disconnect {
                room_name: _,
                sender_name,
            } => {
                self.sessions.remove(&sender_name);
                ()
            }
        }

        Ok(())
    }

    async fn handle_message(&mut self, sender_name: String, message: String) -> Result<()> {
        if message.is_empty() {
            return Err(BlankMessageError.into());
        }

        let chat_message =
            ChatMessage::new(sender_name, message, self.name.to_string(), Zoned::now());

        self.messages.push(chat_message.clone());
        self.broadcast_packet(chat_message.into()).await?;

        Ok(())
    }

    async fn broadcast_packet(&self, packet: ServerPacket) -> Result<()> {
        let _ = join_all(self.sessions.values().map(|session| {
            let packet = packet.clone();
            async move {
                session.send_packet(packet).await.unwrap_or_else(|err| {
                    tracing::trace!(
                        "Error sending packet for session {}: {:?}",
                        session.name,
                        err
                    );
                });
            }
        }))
        .await;

        Ok(())
    }

    async fn send_packet_or_disconnect(
        &mut self,
        message_session: &mut MessageSession,
        server_packet: ServerPacket,
    ) -> Result<()> {
        match message_session.send_packet(server_packet).await {
            Ok(_) => Ok(()),
            Err(_) => {
                self.handle_leave(message_session.name.clone()).await?;
                Ok(())
            }
        }
    }

    async fn send_message_history(&mut self, message_session: &mut MessageSession) -> Result<()> {
        let mut messages = self.messages.clone();
        messages.sort_by(|a, b| a.sent_at.cmp(&b.sent_at));

        for message in messages {
            self.send_packet_or_disconnect(message_session, message.into())
                .await?;
        }
        Ok(())
    }

    async fn notify_join(&mut self, message_session: &MessageSession) -> Result<()> {
        let join_packet = ServerPacket::create_user_joined(
            self.name.to_string(),
            message_session.name.to_string(),
            Zoned::now(),
        );

        self.broadcast_packet(join_packet).await?;
        Ok(())
    }

    pub fn session_count(&self) -> usize {
        self.sessions.len()
    }

    pub fn user_names(&self) -> Vec<String> {
        self.sessions.keys().cloned().collect()
    }
}
